### ISSUE TITLE
Underscore in client URL causes "hostname can't be null" exception

### ISSUE THREAD

**Issue by danielplappert** [micronaut-projects/micronaut-core#1806]
We are deploying some Micronaut Microservices in a docker swarm. In a docker swarm the convention for naming a service is {STACK_NAME}_{SERVICE_NAME}. This means that services have an underscore in their name. When we use the `@Client` annotation to access a service, we get an "hostname can't be null" exception, because the given URL is parsed by URI class and URI class does not allow underscores in hostnames. But in this case, it's a valid URL. 

#### Example:

```
@Singleton
public class ClientExample {

    private final RxHttpClient httpClient;

    public ClientExample(@Client("${api.url}") RxHttpClient httpClient) {
        this.httpClient = httpClient;
    }

    public void doExecute() {
        HttpRequest<?> httpRequest = HttpRequest.GET("/");
        httpClient.exchange(httpRequest).blockingFirst();
    }

}
```

And the application.yml:

```
micronaut:
  application:
    name: uriproblem

api:
  url: http://dockerstack_dockerservice
```

You can download an example application here:
https://github.com/danielplappert/micronaut-uriproblem

#### Stacktrace

```
Caused by: java.lang.IllegalArgumentException: hostname can't be null
	at java.base/java.net.InetSocketAddress.checkHost(InetSocketAddress.java:149)
	at java.base/java.net.InetSocketAddress.createUnresolved(InetSocketAddress.java:254)
	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:126)
	at io.micronaut.http.client.DefaultHttpClient.doConnect(DefaultHttpClient.java:1235)
	at io.micronaut.http.client.DefaultHttpClient.doConnect(DefaultHttpClient.java:1190)
	at io.micronaut.http.client.DefaultHttpClient.doConnect(DefaultHttpClient.java:1163)
	at io.micronaut.http.client.DefaultHttpClient.lambda$null$25(DefaultHttpClient.java:1017)
	at io.reactivex.internal.operators.flowable.FlowableCreate.subscribeActual(FlowableCreate.java:71)
```

### Expected Behaviour

A connection to our service (which is possible without the underscore)

### Actual Behaviour

IllegalArgumentException: hostname can't be null

### Environment Information

- **Operating System**: MacOS 10.14
- **Micronaut Version:** 1.1.3
- **JDK Version:** 8u212, 11.0.1

### Example Application

- https://github.com/danielplappert/micronaut-uriproblem

**Comment by haris-zynka** [comment]
When can we expect this to be resolved as it's still a problem in 1.2.2?

**Comment by graemerocher** [comment]
Thanks for the bump. Rescheduled for 1.2.4
